lib/ui/dart_ui.cc

```diff
@@ -33,64 +33,239 @@
+  V(SceneBuilder, pushOpacity, 6)                      \
+  V(SceneBuilder, pushPhysicalShape, 8)                \
+  V(SceneBuilder, pushShaderMask, 10)                  \
+  V(SceneBuilder, pushTransformHandle, 4)              \
+  V(SceneBuilder, setCheckerboardOffscreenLayers, 2)   \
+  V(SceneBuilder, setCheckerboardRasterCacheImages, 2) \
+  V(SceneBuilder, setRasterizerTracingThreshold, 2)    \
+  V(Scene, dispose, 1)                                 \
+  V(Scene, toImage, 4)                                 \
+  V(SemanticsUpdateBuilder, build, 2)                  \
+  V(SemanticsUpdateBuilder, updateCustomAction, 5)     \
+  V(SemanticsUpdateBuilder, updateNode, 36)            \
+  V(SemanticsUpdate, dispose, 1)
 
-}  // namespace
+#define FFI_FUNCTION_MATCH(FUNCTION, ARGS)                                 \
+  if (strcmp(name, #FUNCTION) == 0 && args == ARGS) {                      \
+    return reinterpret_cast<void*>(                                        \
+        tonic::FfiDispatcher<void, decltype(&FUNCTION), &FUNCTION>::Call); \
+  }
 
-void DartUI::InitForGlobal() {
-  if (!g_natives) {
-    g_natives = new tonic::DartLibraryNatives();
-    Canvas::RegisterNatives(g_natives);
-    CanvasGradient::RegisterNatives(g_natives);
-    CanvasImage::RegisterNatives(g_natives);
-    CanvasPath::RegisterNatives(g_natives);
-    CanvasPathMeasure::RegisterNatives(g_natives);
-    Codec::RegisterNatives(g_natives);
-    ColorFilter::RegisterNatives(g_natives);
-    DartRuntimeHooks::RegisterNatives(g_natives);
-    EngineLayer::RegisterNatives(g_natives);
-    FontCollection::RegisterNatives(g_natives);
-    FragmentProgram::RegisterNatives(g_natives);
-    ImageDescriptor::RegisterNatives(g_natives);
-    ImageFilter::RegisterNatives(g_natives);
-    ImageShader::RegisterNatives(g_natives);
-    ImmutableBuffer::RegisterNatives(g_natives);
-    IsolateNameServerNatives::RegisterNatives(g_natives);
-    NativeStringAttribute::RegisterNatives(g_natives);
-    Paragraph::RegisterNatives(g_natives);
-    ParagraphBuilder::RegisterNatives(g_natives);
-    Picture::RegisterNatives(g_natives);
-    PictureRecorder::RegisterNatives(g_natives);
-    Scene::RegisterNatives(g_natives);
-    SceneBuilder::RegisterNatives(g_natives);
-    SemanticsUpdate::RegisterNatives(g_natives);
-    SemanticsUpdateBuilder::RegisterNatives(g_natives);
-    Vertices::RegisterNatives(g_natives);
-    PlatformConfiguration::RegisterNatives(g_natives);
+#define FFI_METHOD_MATCH(CLASS, METHOD, ARGS)                   \
+  if (strcmp(name, #CLASS "::" #METHOD) == 0 && args == ARGS) { \
+    return reinterpret_cast<void*>(                             \
+        tonic::FfiDispatcher<CLASS, decltype(&CLASS::METHOD),   \
+                             &CLASS::METHOD>::Call);            \
   }
+
+void* ResolveFfiNativeFunction(const char* name, uintptr_t args) {
+  FFI_FUNCTION_LIST(FFI_FUNCTION_MATCH)
+  FFI_METHOD_LIST(FFI_METHOD_MATCH)
+  return nullptr;
```

Comment:
This is replacing what used to be a map lookup with a large list of function calls that boil down to strcmp.

In particular, the order of those calls don't necessarily help, and it's probably not possible to have an ideal ordering of these.

Is there some good reason to not have these in a map that we initialize at startup and then can do faster lookups on?

Reply:
Specifically, I'm a little concerned that some frame that calls `SemanticsUpdate.dispose` 50 times will do over a thousand string comparisons where it used to do 50 map lookups.

Reply:
You're right, the lookup of name -> function itself can certainly be faster than the current O(N). But in practice each binding is only ever looked up once, as the function pointer is stored in Dart as a static variable that is initialised when the binding is evaluated:
```dart
@FfiNative<Void Function()>('bop')
// /\- The above is transformed into the below synthetic code -\/
static final _bop$FfiNative$ptr =
    Pointer<NativeFunction<Void Function()>>
    .fromAddress(_ffi_resolver('...', 'bop', 0))
    .asFunction<void Function()>(isLeaf:false);
```

I'll admit I haven't crunched the numbers on whether it's worth it to trade off the up-front work and memory of the map, versus doing the linear string scans once per function, but my gut feeling is that we're not going to see much practical difference.

I would however be happy to do the rewrite if it concerns you?

Reply:
Ahh I see.  I'd be interested in knowing how this impacts startup time in that case.

Reply:
Does Flutter have some standard benchmark to measure startup time?

Reply:
There are a few devicelab ones with startup in the name

Reply:
I tried some experiments on a low-end device, and based on that I'd prefer to use a map lookup.

I patched `DartUI::InitForIsolate` to insert the FFI dispatchers into a `std::unordered_map<std::string_view, void*>`.

Construction of the map took 220 usecs, and lookups of function names took 1-5 usecs.

By comparison, the strcmp-based lookups ranged from 1-45 usecs, with an average of 9.

I think the upfront cost of the map is a better tradeoff versus the variance of a strcmp lookup that may occur during a frame's critical path.

Also, the map initialization can be done once during the process' first launch of an engine and then retained for when the engine is recreated during an app stop/restart cycle.  The strcmp-style lookups will need to be repeated each time the engine is restarted.